File: java/src/main/java/com/pulumi/cloud/sdk/ApiClientConfiguration.java
// Copyright 2026, Pulumi Corporation.  All rights reserved.

package com.pulumi.cloud.sdk;

import java.time.Duration;
import java.util.function.Supplier;

/**
 * Connection settings shared by every generated {@code *Api} class.
 *
 * <p>{@code host} is the API origin (no {@code /api} suffix — the generated
 * resource paths already start with {@code /api/...}). {@code accessToken}
 * supplies the Pulumi access token sent as {@code Authorization: token <token>};
 * it is a {@link Supplier} so a rotating token can be re-read per request. Return
 * {@code null} for unauthenticated requests.
 */
public class ApiClientConfiguration {
    private String host = "https://api.pulumi.com";
    private Supplier<String> accessToken = () -> null;
    private String source = "pulumi-cloud-sdk";
    private Duration timeout = Duration.ofSeconds(60);

    public String getHost() {
        return host;
    }

    public ApiClientConfiguration setHost(String host) {
        this.host = host;
        return this;
    }

    public Supplier<String> getAccessToken() {
        return accessToken;
    }

    /** Set a static access token (sent as {@code Authorization: token <token>}). */
    public ApiClientConfiguration setAccessToken(String accessToken) {
        this.accessToken = () -> accessToken;
        return this;
    }

    /** Set a dynamic access-token supplier, re-read on every request. */
    public ApiClientConfiguration setAccessToken(Supplier<String> accessToken) {
        this.accessToken = accessToken != null ? accessToken : () -> null;
        return this;
    }

    public String getSource() {
        return source;
    }

    public ApiClientConfiguration setSource(String source) {
        this.source = source;
        return this;
    }

    public Duration getTimeout() {
        return timeout;
    }

    public ApiClientConfiguration setTimeout(Duration timeout) {
        this.timeout = timeout;
        return this;
    }
}

File: dotnet/ApiClientConfiguration.cs
// Copyright 2026, Pulumi Corporation.  All rights reserved.

using System;

namespace Pulumi.Cloud.Sdk
{
    /// <summary>
    /// Connection settings shared by every generated <c>*Api</c> class.
    ///
    /// <para><see cref="Host"/> is the API origin (no <c>/api</c> suffix — the
    /// generated resource paths already start with <c>/api/...</c>).
    /// <see cref="AccessToken"/> supplies the Pulumi access token sent as
    /// <c>Authorization: token &lt;token&gt;</c>; it is a delegate so a rotating
    /// token can be re-read per request. Return <c>null</c> for unauthenticated
    /// requests.</para>
    /// </summary>
    public sealed class ApiClientConfiguration
    {
        public string Host { get; set; } = "https://api.pulumi.com";

        public Func<string> AccessToken { get; set; } = () => null;

        public string Source { get; set; } = "pulumi-cloud-sdk";

        public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(60);

        /// <summary>Set a static access token (sent as <c>Authorization: token &lt;token&gt;</c>).</summary>
        public ApiClientConfiguration WithAccessToken(string token)
        {
            AccessToken = () => token;
            return this;
        }

        /// <summary>Set a dynamic access-token supplier, re-read on every request.</summary>
        public ApiClientConfiguration WithAccessToken(Func<string> token)
        {
            AccessToken = token ?? (() => null);
            return this;
        }
    }
}
